File: scripts/_validate_dynamic_stack.mjs
// Standalone validation for the dynamic Sheets order-row formatter.
//
// Run from the repo root with:
//     node scripts/_validate_dynamic_stack.mjs
//
// Re-implements `buildOrderRow` and `buildThreeSlotRow` from
// `lib/webhooks/google-sheets.ts` inline so the script has zero
// TS/build dependencies. The two implementations MUST stay in sync —
// this script exists specifically so production regressions like the
// "2/1/3 vs expected 2/1/1/3" multi-upsell truncation can never come
// back undetected.
//
// Coverage:
//   • Arbitrary product counts (1, 2, 4, 6, 9 segments).
//   • Mixed upsell + cross-sell stacking.
//   • Repeated quantities (1..7).
//   • Duplicated SKUs.
//   • Insertion-order preservation within each bucket.
//   • Source ordering (base → upsell → cross_sell) regardless of input order.
//   • Per-cell alignment (sku / productName / totalQuantity / productUrl).
//   • Empty input.
//   • Unknown / missing source defaults to "base".
//   • The legacy `buildThreeSlotRow` shim now also returns one segment
//     per line (no fixed-slot padding) — guards against any caller
//     accidentally rolling back to the broken 3-slot truncation.

const SOURCE_RANK = { base: 0, upsell: 1, cross_sell: 2 };

function buildOrderRow(lines) {
  const ordered = lines
    .map((line, idx) => ({
      line: {
        ...line,
        source:
          line.source === "base" ||
          line.source === "upsell" ||
          line.source === "cross_sell"
            ? line.source
            : "base",
      },
      idx,
    }))
    .sort((a, b) => {
      const r = SOURCE_RANK[a.line.source] - SOURCE_RANK[b.line.source];
      return r !== 0 ? r : a.idx - b.idx;
    })
    .map(({ line }) => line);

  return {
    sku: ordered.map((l) => l.sku ?? "").join("/"),
    productName: ordered.map((l) => l.name ?? "").join("/"),
    totalQuantity: ordered.map((l) => String(l.quantity ?? 0)).join("/"),
    productUrl: ordered.map((l) => l.url ?? "").join("/"),
  };
}

function buildThreeSlotRow(lines) {
  const dyn = buildOrderRow(
    lines.map((l) => ({
      sku: l.sku,
      name: l.name,
      quantity: l.quantity,
      url: "",
      source: l.source,
    })),
  );
  return {
    sku: dyn.sku,
    productName: dyn.productName,
    totalQuantity: dyn.totalQuantity,
  };
}

let passed = 0;
let failed = 0;
function expect(label, actual, expected) {
  const ok = JSON.stringify(actual) === JSON.stringify(expected);
  if (ok) {
    passed++;
    console.log(`  \u2713 ${label}`);
  } else {
    failed++;
    console.log(`  \u2717 ${label}`);
    console.log(`      expected: ${JSON.stringify(expected)}`);
    console.log(`      actual:   ${JSON.stringify(actual)}`);
  }
}

const url = (slug) => `https://elfanaa.com/${slug}`;
const sku = (n) => `FN-${String(n).padStart(3, "0")}`;

console.log("\nbuildOrderRow — single product (1 segment)");

expect(
  "Sugarbear x3 only  →  '3'",
  buildOrderRow([
    { sku: "FN-SUG-004", name: "سوجاربير", quantity: 3, url: url("sugarbear"), source: "base" },
  ]),
  {
    sku: "FN-SUG-004",
    productName: "سوجاربير",
    totalQuantity: "3",
    productUrl: url("sugarbear"),
  },
);

console.log("\nbuildOrderRow — base + cross-sell (2 segments)");

expect(
  "Sugarbear x1 + cross-sell x3  →  '1/3' (2 segments, no padding)",
  buildOrderRow([
    { sku: "FN-SUG-004", name: "سوجاربير", quantity: 1, url: url("sugarbear"), source: "base" },
    { sku: "FN-CREAM-002", name: "كريم", quantity: 3, url: url("cream"), source: "cross_sell" },
  ]),
  {
    sku: "FN-SUG-004/FN-CREAM-002",
    productName: "سوجاربير/كريم",
    totalQuantity: "1/3",
    productUrl: `${url("sugarbear")}/${url("cream")}`,
  },
);

console.log("\nbuildOrderRow — the exact multi-upsell regression scenario");

expect(
  "Base x2 + upsell #1 + upsell #2 + cross x3  →  '2/1/1/3' (the user's expected output)",
  buildOrderRow([
    { sku: "FN-SUG-004", name: "Sugarbear", quantity: 2, url: url("sugarbear"), source: "base" },
    { sku: "FN-UPS-99-A", name: "Upsell A", quantity: 1, url: url("upsell-a"), source: "upsell" },
    { sku: "FN-UPS-99-B", name: "Upsell B", quantity: 1, url: url("upsell-b"), source: "upsell" },
    { sku: "FN-CRS-001", name: "Cross", quantity: 3, url: url("cross"), source: "cross_sell" },
  ]),
  {
    sku: "FN-SUG-004/FN-UPS-99-A/FN-UPS-99-B/FN-CRS-001",
    productName: "Sugarbear/Upsell A/Upsell B/Cross",
    totalQuantity: "2/1/1/3",
    productUrl: `${url("sugarbear")}/${url("upsell-a")}/${url("upsell-b")}/${url("cross")}`,
  },
);

console.log("\nbuildOrderRow — large funnel: base + 5 upsells (6 segments)");

expect(
  "Base + 5 upsells  →  '1/1/1/1/1/1'  +  6 SKU segments",
  buildOrderRow([
    { sku: "B", name: "Base", quantity: 1, url: "u0", source: "base" },
    { sku: "U1", name: "U1", quantity: 1, url: "u1", source: "upsell" },
    { sku: "U2", name: "U2", quantity: 1, url: "u2", source: "upsell" },
    { sku: "U3", name: "U3", quantity: 1, url: "u3", source: "upsell" },
    { sku: "U4", name: "U4", quantity: 1, url: "u4", source: "upsell" },
    { sku: "U5", name: "U5", quantity: 1, url: "u5", source: "upsell" },
  ]),
  {
    sku: "B/U1/U2/U3/U4/U5",
    productName: "Base/U1/U2/U3/U4/U5",
    totalQuantity: "1/1/1/1/1/1",
    productUrl: "u0/u1/u2/u3/u4/u5",
  },
);

console.log("\nbuildOrderRow — base + 7 cross-sells (8 segments)");

expect(
  "Base + 7 cross-sells  →  8 segments, deterministic order",
  buildOrderRow([
    { sku: "B", name: "B", quantity: 1, url: "ub", source: "base" },
    { sku: "C1", name: "C1", quantity: 1, url: "u1", source: "cross_sell" },
    { sku: "C2", name: "C2", quantity: 2, url: "u2", source: "cross_sell" },
    { sku: "C3", name: "C3", quantity: 3, url: "u3", source: "cross_sell" },
    { sku: "C4", name: "C4", quantity: 4, url: "u4", source: "cross_sell" },
    { sku: "C5", name: "C5", quantity: 5, url: "u5", source: "cross_sell" },
    { sku: "C6", name: "C6", quantity: 6, url: "u6", source: "cross_sell" },
    { sku: "C7", name: "C7", quantity: 7, url: "u7", source: "cross_sell" },
  ]),
  {
    sku: "B/C1/C2/C3/C4/C5/C6/C7",
    productName: "B/C1/C2/C3/C4/C5/C6/C7",
    totalQuantity: "1/1/2/3/4/5/6/7",
    productUrl: "ub/u1/u2/u3/u4/u5/u6/u7",
  },
);

console.log("\nbuildOrderRow — base + 9 mixed extras (10 segments)");

expect(
  "Base + 4 upsells + 5 cross-sells (mixed)  →  10 segments, base→upsell→cross_sell",
  buildOrderRow([
    { sku: "B", name: "B", quantity: 1, url: "ub", source: "base" },
    { sku: "C1", name: "C1", quantity: 1, url: "uc1", source: "cross_sell" },
    { sku: "U1", name: "U1", quantity: 1, url: "uu1", source: "upsell" },
    { sku: "C2", name: "C2", quantity: 1, url: "uc2", source: "cross_sell" },
    { sku: "U2", name: "U2", quantity: 1, url: "uu2", source: "upsell" },
    { sku: "C3", name: "C3", quantity: 1, url: "uc3", source: "cross_sell" },
    { sku: "U3", name: "U3", quantity: 1, url: "uu3", source: "upsell" },
    { sku: "C4", name: "C4", quantity: 1, url: "uc4", source: "cross_sell" },
    { sku: "U4", name: "U4", quantity: 1, url: "uu4", source: "upsell" },
    { sku: "C5", name: "C5", quantity: 1, url: "uc5", source: "cross_sell" },
  ]),
  {
    sku: "B/U1/U2/U3/U4/C1/C2/C3/C4/C5",
    productName: "B/U1/U2/U3/U4/C1/C2/C3/C4/C5",
    totalQuantity: "1/1/1/1/1/1/1/1/1/1",
    productUrl: "ub/uu1/uu2/uu3/uu4/uc1/uc2/uc3/uc4/uc5",
  },
);

console.log("\nbuildOrderRow — duplicate SKUs (every line preserved)");

expect(
  "Same SKU added as base + cross-sell  →  both segments emit (no merge)",
  buildOrderRow([
    { sku: "FN-DUP", name: "Dup", quantity: 2, url: "u", source: "base" },
    { sku: "FN-DUP", name: "Dup", quantity: 1, url: "u", source: "cross_sell" },
  ]),
  {
    sku: "FN-DUP/FN-DUP",
    productName: "Dup/Dup",
    totalQuantity: "2/1",
    productUrl: "u/u",
  },
);

console.log("\nbuildOrderRow — insertion order preserved within each bucket");

expect(
  "Upsells UA, UB, UC kept in insertion order regardless of source input order",
  buildOrderRow([
    { sku: "UA", name: "UA", quantity: 1, url: "a", source: "upsell" },
    { sku: "B",  name: "B",  quantity: 1, url: "b", source: "base" },
    { sku: "UB", name: "UB", quantity: 1, url: "ub", source: "upsell" },
    { sku: "UC", name: "UC", quantity: 1, url: "uc", source: "upsell" },
  ]),
  {
    // base first (B), then upsells in their insertion order (UA, UB, UC)
    sku: "B/UA/UB/UC",
    productName: "B/UA/UB/UC",
    totalQuantity: "1/1/1/1",
    productUrl: "b/a/ub/uc",
  },
);

console.log("\nbuildOrderRow — edge cases");

expect(
  "Empty input  →  every field is the empty string",
  buildOrderRow([]),
  { sku: "", productName: "", totalQuantity: "", productUrl: "" },
);

expect(
  "Unknown source collapses to base",
  buildOrderRow([
    { sku: "X", name: "X", quantity: 1, url: "ux", source: "shipping" },
  ]),
  { sku: "X", productName: "X", totalQuantity: "1", productUrl: "ux" },
);

expect(
  "Missing per-line url  →  empty string segment (alignment preserved)",
  buildOrderRow([
    { sku: "A", name: "A", quantity: 1, source: "base" },
    { sku: "B", name: "B", quantity: 1, source: "upsell" },
  ]),
  { sku: "A/B", productName: "A/B", totalQuantity: "1/1", productUrl: "/" },
);

expect(
  "Missing quantity defaults to 0 in the joined string",
  buildOrderRow([
    { sku: "X", name: "X", url: "u", source: "base" },
  ]),
  { sku: "X", productName: "X", totalQuantity: "0", productUrl: "u" },
);

console.log("\nbuildOrderRow — large quantities + many SKUs alignment");

const big = Array.from({ length: 12 }, (_, i) => ({
  sku: sku(i + 1),
  name: `P${i + 1}`,
  quantity: ((i * 3) % 7) + 1,
  url: `u${i + 1}`,
  source: i === 0 ? "base" : i % 2 === 0 ? "upsell" : "cross_sell",
}));
const bigOut = buildOrderRow(big);
expect(
  "12 segments — sku / name / qty / url cells all have exactly 12 segments",
  {
    skuSeg: bigOut.sku.split("/").length,
    nameSeg: bigOut.productName.split("/").length,
    qtySeg: bigOut.totalQuantity.split("/").length,
    urlSeg: bigOut.productUrl.split("/").length,
  },
  { skuSeg: 12, nameSeg: 12, qtySeg: 12, urlSeg: 12 },
);

console.log("\nbuildThreeSlotRow shim — now dynamic (no fixed-slot truncation)");

expect(
  "Legacy shim: base x2 + upsell + upsell + cross x3  →  '2/1/1/3' (no truncation)",
  buildThreeSlotRow([
    { sku: "B", name: "B", quantity: 2, source: "base" },
    { sku: "U1", name: "U1", quantity: 1, source: "upsell" },
    { sku: "U2", name: "U2", quantity: 1, source: "upsell" },
    { sku: "C", name: "C", quantity: 3, source: "cross_sell" },
  ]),
  { sku: "B/U1/U2/C", productName: "B/U1/U2/C", totalQuantity: "2/1/1/3" },
);

expect(
  "Legacy shim: single base x3  →  '3' (no padding to 3 slots)",
  buildThreeSlotRow([{ sku: "X", name: "X", quantity: 3, source: "base" }]),
  { sku: "X", productName: "X", totalQuantity: "3" },
);

console.log("");
console.log(`Result: ${passed} passed, ${failed} failed`);
process.exit(failed === 0 ? 0 : 1);

File: backend/_validate_dynamic_stack.py
"""Standalone validation for the dynamic Sheets order-row formatter (Python).

Run from the repo root with:
    python backend/_validate_dynamic_stack.py

Mirrors `scripts/_validate_dynamic_stack.mjs` so the two backends
(Next.js fallback + FastAPI two-tier) emit IDENTICAL Sheets rows for
the same inputs. The dynamic builder is the new (post-2026-05-19)
contract that supports unbounded multi-upsell / multi-cross-sell
stacks; this script exists specifically so the live "2/1/3 vs
expected 2/1/1/3" regression can never come back undetected.
"""

from __future__ import annotations

import sys
from pathlib import Path

# Reconfigure stdout for Windows consoles that default to cp1252; the
# regression labels contain Arabic.
try:
    sys.stdout.reconfigure(encoding="utf-8")  # type: ignore[attr-defined]
except Exception:
    pass

ROOT = Path(__file__).resolve().parent
sys.path.insert(0, str(ROOT))

from app.services.webhooks import (
    build_order_row,
    build_sheets_order_row,
    build_sheets_order_update_row,
    build_three_slot_row,
    compose_full_address,
)


passed = 0
failed = 0


def expect(label: str, actual, expected) -> None:
    global passed, failed
    if actual == expected:
        passed += 1
        print(f"  \u2713 {label}")
    else:
        failed += 1
        print(f"  \u2717 {label}")
        print(f"      expected: {expected!r}")
        print(f"      actual:   {actual!r}")


def line(
    *,
    sku: str,
    name: str,
    quantity: int,
    source: str = "base",
    url: str = "",
):
    return {
        "sku": sku,
        "name": name,
        "quantity": quantity,
        "url": url,
        "source": source,
    }


# ── build_order_row ─────────────────────────────────────────────────────────

print("\nbuild_order_row — single product (1 segment)")
expect(
    "Sugarbear x3 only  →  '3'",
    build_order_row(
        [
            line(
                sku="FN-SUG-004",
                name="سوجاربير",
                quantity=3,
                source="base",
                url="https://elfanaa.com/sugarbear",
            )
        ]
    ),
    {
        "sku": "FN-SUG-004",
        "productName": "سوجاربير",
        "totalQuantity": "3",
        "productUrl": "https://elfanaa.com/sugarbear",
    },
)

print("\nbuild_order_row — base + cross-sell (2 segments)")
expect(
    "Sugarbear x1 + cross-sell x3  →  '1/3' (2 segments, no padding)",
    build_order_row(
        [
            line(sku="FN-SUG-004", name="سوجاربير", quantity=1, url="u-s", source="base"),
            line(sku="FN-CREAM-002", name="كريم", quantity=3, url="u-c", source="cross_sell"),
        ]
    ),
    {
        "sku": "FN-SUG-004/FN-CREAM-002",
        "productName": "سوجاربير/كريم",
        "totalQuantity": "1/3",
        "productUrl": "u-s/u-c",
    },
)

print("\nbuild_order_row — exact multi-upsell regression")
expect(
    "Base x2 + upsell #1 + upsell #2 + cross x3  →  '2/1/1/3' (the user's expected output)",
    build_order_row(
        [
            line(sku="FN-SUG-004", name="Sugarbear", quantity=2, url="us", source="base"),
            line(sku="FN-UPS-99-A", name="Upsell A", quantity=1, url="ua", source="upsell"),
            line(sku="FN-UPS-99-B", name="Upsell B", quantity=1, url="ub", source="upsell"),
            line(sku="FN-CRS-001", name="Cross", quantity=3, url="uc", source="cross_sell"),
        ]
    ),
    {
        "sku": "FN-SUG-004/FN-UPS-99-A/FN-UPS-99-B/FN-CRS-001",
        "productName": "Sugarbear/Upsell A/Upsell B/Cross",
        "totalQuantity": "2/1/1/3",
        "productUrl": "us/ua/ub/uc",
    },
)

print("\nbuild_order_row — large funnel: base + 5 upsells (6 segments)")
expect(
    "Base + 5 upsells  →  6 segments",
    build_order_row(
        [
            line(sku="B", name="Base", quantity=1, url="u0", source="base"),
            line(sku="U1", name="U1", quantity=1, url="u1", source="upsell"),
            line(sku="U2", name="U2", quantity=1, url="u2", source="upsell"),
            line(sku="U3", name="U3", quantity=1, url="u3", source="upsell"),
            line(sku="U4", name="U4", quantity=1, url="u4", source="upsell"),
            line(sku="U5", name="U5", quantity=1, url="u5", source="upsell"),
        ]
    ),
    {
        "sku": "B/U1/U2/U3/U4/U5",
        "productName": "Base/U1/U2/U3/U4/U5",
        "totalQuantity": "1/1/1/1/1/1",
        "productUrl": "u0/u1/u2/u3/u4/u5",
    },
)

print("\nbuild_order_row — base + 7 cross-sells (8 segments)")
expect(
    "Base + 7 cross-sells  →  8 segments with quantities 1..7",
    build_order_row(
        [
            line(sku="B", name="B", quantity=1, url="ub", source="base"),
            line(sku="C1", name="C1", quantity=1, url="u1", source="cross_sell"),
            line(sku="C2", name="C2", quantity=2, url="u2", source="cross_sell"),
            line(sku="C3", name="C3", quantity=3, url="u3", source="cross_sell"),
            line(sku="C4", name="C4", quantity=4, url="u4", source="cross_sell"),
            line(sku="C5", name="C5", quantity=5, url="u5", source="cross_sell"),
            line(sku="C6", name="C6", quantity=6, url="u6", source="cross_sell"),
            line(sku="C7", name="C7", quantity=7, url="u7", source="cross_sell"),
        ]
    ),
    {
        "sku": "B/C1/C2/C3/C4/C5/C6/C7",
        "productName": "B/C1/C2/C3/C4/C5/C6/C7",
        "totalQuantity": "1/1/2/3/4/5/6/7",
        "productUrl": "ub/u1/u2/u3/u4/u5/u6/u7",
    },
)

print("\nbuild_order_row — duplicate SKUs (each line preserved)")
expect(
    "Same SKU added as base + cross-sell  →  both segments emitted, no merge",
    build_order_row(
        [
            line(sku="FN-DUP", name="Dup", quantity=2, url="u", source="base"),
            line(sku="FN-DUP", name="Dup", quantity=1, url="u", source="cross_sell"),
        ]
    ),
    {
        "sku": "FN-DUP/FN-DUP",
        "productName": "Dup/Dup",
        "totalQuantity": "2/1",
        "productUrl": "u/u",
    },
)

print("\nbuild_order_row — source ordering (base → upsell → cross_sell)")
expect(
    "Mixed input order: upsell first, base second, cross third  →  reordered correctly",
    build_order_row(
        [
            line(sku="UA", name="UA", quantity=1, url="a", source="upsell"),
            line(sku="B",  name="B",  quantity=1, url="b", source="base"),
            line(sku="C",  name="C",  quantity=1, url="c", source="cross_sell"),
            line(sku="UB", name="UB", quantity=1, url="ub", source="upsell"),
        ]
    ),
    {
        "sku": "B/UA/UB/C",
        "productName": "B/UA/UB/C",
        "totalQuantity": "1/1/1/1",
        "productUrl": "b/a/ub/c",
    },
)

print("\nbuild_order_row — edge cases")
expect(
    "Empty list  →  every field is the empty string",
    build_order_row([]),
    {"sku": "", "productName": "", "totalQuantity": "", "productUrl": ""},
)
expect(
    "Unknown source collapses to base",
    build_order_row(
        [line(sku="X", name="X", quantity=1, url="u", source="shipping")]
    ),
    {"sku": "X", "productName": "X", "totalQuantity": "1", "productUrl": "u"},
)
expect(
    "Missing URL  →  empty-string segment (alignment preserved)",
    build_order_row(
        [
            {"sku": "A", "name": "A", "quantity": 1, "source": "base"},
            {"sku": "B", "name": "B", "quantity": 1, "source": "upsell"},
        ]
    ),
    {"sku": "A/B", "productName": "A/B", "totalQuantity": "1/1", "productUrl": "/"},
)

# ── build_sheets_order_row uses build_order_row internally ──────────────────

print("\nbuild_sheets_order_row — multi-segment row preserves full state")
sheets = build_sheets_order_row(
    order_id="ord_x",
    order_date_ksa="19/05/2026",
    full_name="X",
    phone_digits="966500000000",
    full_address="",
    items=[
        line(sku="A", name="A", quantity=1, url="ua", source="base"),
        line(sku="B", name="B", quantity=2, url="ub", source="cross_sell"),
        line(sku="C", name="C", quantity=3, url="uc", source="cross_sell"),
    ],
    total_minor=49700,
    currency="SAR",
    fallback_product_url="https://elfanaa.com/sugarbear",
)
expect(
    "order row contains every accepted line's segment for SKU/name/qty/url",
    {
        "sku": sheets["sku"],
        "name": sheets["productName"],
        "qty": sheets["totalQuantity"],
        "url": sheets["productUrl"],
        "variantPrice": sheets["variantPrice"],
        "kind": sheets["kind"],
    },
    {
        "sku": "A/B/C",
        "name": "A/B/C",
        "qty": "1/2/3",
        "url": "ua/ub/uc",
        "variantPrice": 497,
        "kind": "order",
    },
)

print("\nbuild_sheets_order_row — fallback URL used only when EVERY line has no URL")
sheets_fallback = build_sheets_order_row(
    order_id="ord_y",
    order_date_ksa="19/05/2026",
    full_name="Y",
    phone_digits="966500000000",
    full_address="",
    items=[
        line(sku="A", name="A", quantity=1, url="", source="base"),
        line(sku="B", name="B", quantity=2, url="", source="cross_sell"),
    ],
    total_minor=39800,
    currency="SAR",
    fallback_product_url="https://elfanaa.com/sugarbear",
)
expect(
    "All lines URL-less  →  productUrl falls back to the request referer",
    sheets_fallback["productUrl"],
    "https://elfanaa.com/sugarbear",
)

# ── build_sheets_order_update_row (the multi-upsell rewrite path) ───────────

print("\nbuild_sheets_order_update_row — full-state rewrite carries every line")
update = build_sheets_order_update_row(
    order_id="ord_z",
    items=[
        line(sku="B", name="B", quantity=2, url="ub", source="base"),
        line(sku="U1", name="U1", quantity=1, url="uu1", source="upsell"),
        line(sku="U2", name="U2", quantity=1, url="uu2", source="upsell"),
        line(sku="C", name="C", quantity=3, url="uc", source="cross_sell"),
    ],
    total_minor=89600,
    currency="SAR",
)
expect(
    "order_update payload — kind + every segment + variant price all present",
    update,
    {
        "kind": "order_update",
        "orderId": "ord_z",
        "sku": "B/U1/U2/C",
        "productName": "B/U1/U2/C",
        "totalQuantity": "2/1/1/3",
        "productUrl": "ub/uu1/uu2/uc",
        "variantPrice": 896,
        "currency": "SAR",
    },
)

# ── build_three_slot_row shim — must now be dynamic (no fixed-slot padding) ─

print("\nbuild_three_slot_row shim — dynamic, no truncation")
expect(
    "Legacy shim: base x2 + 2 upsells + cross x3  →  '2/1/1/3' (no truncation)",
    build_three_slot_row(
        [
            {"sku": "B", "name": "B", "quantity": 2, "source": "base"},
            {"sku": "U1", "name": "U1", "quantity": 1, "source": "upsell"},
            {"sku": "U2", "name": "U2", "quantity": 1, "source": "upsell"},
            {"sku": "C", "name": "C", "quantity": 3, "source": "cross_sell"},
        ]
    ),
    {"sku": "B/U1/U2/C", "productName": "B/U1/U2/C", "totalQuantity": "2/1/1/3"},
)
expect(
    "Legacy shim: single base x3  →  '3' (no padding to 3 slots)",
    build_three_slot_row(
        [{"sku": "X", "name": "X", "quantity": 3, "source": "base"}]
    ),
    {"sku": "X", "productName": "X", "totalQuantity": "3"},
)

# ── compose_full_address — unchanged ────────────────────────────────────────

print("\ncompose_full_address — unchanged contract")
expect("City only", compose_full_address("الرياض", None), "الرياض")
expect("Address only", compose_full_address(None, "حي النخيل"), "حي النخيل")
expect("Both", compose_full_address("الرياض", "حي النخيل"), "الرياض — حي النخيل")
expect("Whitespace + whitespace", compose_full_address("   ", "  "), "")
expect("None + None", compose_full_address(None, None), "")

print("")
print(f"Result: {passed} passed, {failed} failed")
sys.exit(0 if failed == 0 else 1)
